misc/python/materialize/data_ingest/definition.py

```diff
@@ -0,0 +1,185 @@
+
+        transactions = []
+        for i in range(count):
+            for field in fields_with_values:
+                if field.is_key:
+                    field.value = field.typ.num_value(self.current_key)
+                else:
+                    field.value = field.typ.random_value(self.record_size)
+            self.current_key += 1
+            transactions.append(
+                Transaction(
+                    [
+                        RowList(
+                            [
+                                Row(
+                                    fields=deepcopy(fields_with_values),
+                                    operation=Operation.INSERT,
+                                )
+                            ]
+                        )
+                    ]
+                )
+            )
+
+        return transactions
+
+
+class Upsert(Definition):
+    def __init__(self, keyspace: Keyspace, count: Records, record_size: RecordSize):
+        self.keyspace = keyspace
+        self.count = count
+        self.record_size = record_size
+
+    def generate(self, fields: List[Field]) -> List[Transaction]:
+        if self.count == Records.ONE:
+            count = 1
+        elif self.count == Records.MANY:
+            count = 1000
+        else:
+            raise ValueError(f"Unexpected count {self.count}")
+
+        fields_with_values = deepcopy(fields)
+
+        transactions = []
+        for i in range(count):
+            for field in fields_with_values:
+                if field.is_key:
+                    if self.keyspace == Keyspace.SINGLE_VALUE:
+                        field.value = field.typ.num_value(0)
+                    else:
+                        raise NotImplementedError
+                else:
+                    field.value = field.typ.random_value(self.record_size)
+
+            transactions.append(
+                Transaction(
+                    [
+                        RowList(
+                            [
+                                Row(
+                                    fields=deepcopy(fields_with_values),
+                                    operation=Operation.UPSERT,
+                                )
+                            ]
+                        )
+                    ]
+                )
+            )
+
+        return transactions
+
+
+class Delete(Definition):
+    def __init__(
+        self,
+        number_of_records: Records,
+        record_size: RecordSize,
+        num: Optional[int] = None,
+    ):
+        self.number_of_records = number_of_records
+        self.record_size = record_size
+        self.num = num
+
+    def generate(self, fields: List[Field]) -> List[Transaction]:
+        transactions = []
+
+        fields_with_values = [field for field in fields if field.is_key]
+
+        if self.number_of_records == Records.ONE:
+            for field in fields_with_values:
+                field.value = field.typ.random_value(self.record_size)
```

Comment:
Is the value that you are trying to delete known to exist? This may be important as deleting a non-existent value from Postgres is a no-op, but if you have a delete in an `UPSERT` Kafka topic, do we simply ignore it and move on , or fail the source with an error?

A data structure such as https://pypi.org/project/roaringbitmap/ may be useful.

Reply:
Not known. There is no error in Kafka on deleting a non-existing key based on what I can tell.